Framework/Beamline/src/ComponentInfo.cpp

```diff
@@ -359,5 +511,235 @@ bool ComponentInfo::isStructuredBank(const size_t componentIndex) const {
+}
+
+/**
+ * Set the scan interval using nanosecond offsets from 00:00:00 01/01/1990 epoch
+ *
+ * This MUST be called before any merging has happened, therefore index is the
+ *time-independent
+ * component index.
+ *
+ * @param index : component to which to add the scan interval
+ * component
+ * @param interval : Time interval for scan point
+ */
+void ComponentInfo::setScanInterval(
+    const size_t index, const std::pair<int64_t, int64_t> &interval) {
+  // Enforces setting scan intervals BEFORE time indexed positions and rotations
+  checkNoTimeDependence();
+  checkScanInterval(interval);
+  if (m_scanIntervals && m_isSyncScan) {
+    throw std::runtime_error("ComponentInfo has been initialized with a "
+                             "synchonous scan, cannot set scan interval for "
+                             "individual component.");
+  }
+  if (!m_scanIntervals) {
+    m_isSyncScan = false;
+    initScanIntervals();
+  }
+  for (auto subIndex : this->componentsInSubtree(index)) {
+    if (m_detectorInfo && isDetector(subIndex)) {
+      m_detectorInfo->setScanInterval(subIndex, interval);
+    } else {
+      m_scanIntervals.access()[compOffsetIndex(subIndex)] = interval;
+    }
+  }
+}
+
+void ComponentInfo::setScanInterval(
+    const std::pair<int64_t, int64_t> &interval) {
+  // Enforces setting scan intervals BEFORE time indexed positions and rotations
+  checkNoTimeDependence();
+  checkScanInterval(interval);
+  if (m_scanIntervals && !m_isSyncScan) {
+    throw std::runtime_error(
+        "ComponentInfo has been initialized with a "
+        "asynchonous scan, cannot set synchronous scan interval.");
+  }
+  if (!m_scanIntervals) {
+    m_isSyncScan = true;
+    initScanIntervals();
+  }
+  m_scanIntervals.access()[0] = interval;
+  if (m_detectorInfo) {
+    m_detectorInfo->setScanInterval(interval);
+  }
+}
+
+/**
+Merges the contents of other ComponentInfo into this. The assumption is that
+this has no time dependence prior to this operation.
+ *
+ * Scan intervals in both other and this must be set. Intervals must be
+ * identical or non-overlapping. If they are identical all other parameters (for
+ * that index) must match.
+ *
+ * Time indices in `this` are preserved. Time indices added from `other` are
+ * incremented by the scan count of that detector in `this`. The relative order
+ * of time indices added from `other` is preserved. If the interval for a time
+ * index in `other` is identical to a corresponding interval in `this`, it is
+ * ignored, i.e., no time index is added.
+**/
+void ComponentInfo::merge(const ComponentInfo &other) {
+  if (!m_scanCounts)
+    this->initScanCounts();
+  const auto toMerge = buildMergeIndices(other);
+  if (m_isSyncScan)
+    throw std::runtime_error(
+        "Sync scans not currently supported for ComponentInfo");
+  initIndices();
+  auto scanCounts(m_scanCounts);
+  for (size_t linearIndex = 0; linearIndex < other.m_positions->size();
+       ++linearIndex) {
```

Comment:
We had discussed forbidding scanning anything that is not something like a detector bank (in particular source and sample). How is that handled here?